MediaStreamTrack.getFrameStats()

henbos opened the pull request:
Fixes #76.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/henbos/mediacapture-extensions/pull/77.html" title="Last updated on Jan 27, 2023, 1:38 PM UTC (336acb0)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-extensions/77/f885c2a...henbos:336acb0.html" title="Last updated on Jan 27, 2023, 1:38 PM UTC (336acb0)">Diff</a>

alvestrand commented on index.html:
```diff
@@ -418,6 +419,49 @@ <h2>Transferable MediaStreamTrack</h2>
       <p>The underlying source is supposed to be kept alive between the transfer and transfer-receiving steps, or as long as the data holder is alive.
       In a sense, between these steps, the data holder is attached to the underlying source as if it was a track.</p>
     </div>
+    <h3>MediaStreamTrack frame counters</h3>
```
Add a few words about why frame counters are nice to have? (Like: Frame counters allow you to tell why the frame rate experienced is different from the expected one - for instance, the camera may be slowing down frame production because it's dark, the settings applied may produce decimation of frames, or the system may be CPU starved and therefore dropping frames.)

henbos replied:
Done!

jan-ivar commented:
This should be converted to draft I think, since this doesn't have any agreement yet.

jan-ivar commented on index.html:
```diff
@@ -418,6 +419,58 @@ <h2>Transferable MediaStreamTrack</h2>
+            Only defined for video tracks. Incremented every time a source frame
+            was dropped prior to being emitted. This happens when a frame is
+            decimated in order to achieve the desired
+            {{MediaTrackSettings/frameRate}} or if the user agent dropped the
+            frame for other reasons, such as if the pipeline is not consuming
+            emitted frames fast enough or if the track does not have any sinks.
```
This doesn't sound like an inherent property of the track anymore.

The idea that frame-rate might depend on consumption seems problematic and potentially user agent dependent. I don't think this has come up before. If the actual frame-rate is determined downstream and depends on downstream, perhaps it is best measured downstream as well? See https://github.com/w3c/mediacapture-extensions/issues/76#issuecomment-1276459896.

cc @Pehrsons @padenot thoughts?

henbos replied:
It might make more sense to count a frame that survives until emission as a framesEmitted even if there are no sinks here, since it's the would-be FPS that any consumer would experience if there were any consumers.

No strong opinion but happy to change this

Pehrsons replied:
I mostly wonder what the definition of "emission" and "pipeline" is.

But I get the gist of this: `framesDropped` would for instance, for a camera that only captures 30fps but with a track setting of 15fps, record those extraneous frames that had to be dropped to reach 15. Or it could record those frames dropped in order to not lag behind under high load, in case frames from a camera go through some processing/conversion step.

I was also thinking `framesDropped` would get pretty obvious with `VideoTrackGenerator`, but the processing happens in `MediaStreamTrackProcessor` and so does not apply. A processor could provide its own measure of frames dropped from its buffer, I guess.

What if a track is disabled? I imagine `framesDropped` would be incrementing at this point in case the source is still capturing? Might need clarification.

As long as `framesDropped` doesn't depend on consumption by track sinks, and to my understanding it doesn't, this seems fine to me.

henbos replied:
I'll create a slide for this for the next WG meeting on Tuesday, but I'm thinking of frames "before" or "after" reaching the media source. A frame produced that doesn't reach the source is "dropped" and a frame produced that does gets "emitted" to the rest of the pipeline. WebRTC for example is later in the pipeline and already has frame counters, but it doesn't know about frames dropped earlier and as such we can't tell the FPS of the camera (unless we assume 0 frame drops)

jan-ivar replied:
What does "reach the source" mean?

henbos replied:
That the frame is ready to be handed off to sinks, if there are any.

We usually think of the track's source as the thing producing the frames, but in order for a frame to go from the camera to something consuming frames there could be several steps along the way.

Examples:
- The camera process receives a frame from the camera (++framesProduced), the frame is passed to the renderer process and eventually ends up where the media source forwards it to its sinks (++framesEmitted).
- The camera process receives a frame from the camera (++framesProduced), the frame is bassed to the renderer process but the renderer process' buffer pool is out of frames so the frame is discarded (++framesDropped).

henbos commented on index.html:
```diff
@@ -418,6 +419,58 @@ <h2>Transferable MediaStreamTrack</h2>
+          </p>
+        </dd>
+        <dt>
+          <dfn>framesDropped</dfn> of type {{unsigned long long}}, readonly
```
We could expose capturedFrames instead of framesDropped, it's easy to calculate one from the other.

I'm interested in the delta between capturedFrames and framesEmitted as a signal that frames are getting lost somewhere. Other than frame decimation, it could be a sign of CPU overload

henbos commented on index.html:
```diff
@@ -418,6 +419,58 @@ <h2>Transferable MediaStreamTrack</h2>
+      <p>The WebIDL changes to add frame counters are the following:
+      <pre class="idl"
+>partial interface MediaStreamTrack {
+  readonly attribute unsigned long long framesEmitted;
+  readonly attribute unsigned long long framesDropped;
```
Sync API sounds good to me

Pehrsons replied:
Not saying it shouldn't be async, but:
A dispatch would be an implementation detail. One could do something with an atomic or a lock and caching the result for the duration of the current task for stability, for instance.

youennf commented:
I think this deserves some WebRTC WG meeting time.

henbos commented:
> I think this deserves some WebRTC WG meeting time.

Ack. I'll join the meeting on Tuesday.

henbos reviewed (commented):
PR updated:
- It's now async, track.getStats().
- The metrics are "framesCaptured" and "framesEmitted".
- Dropped frames can be calculated as "framesCaptured - framesEmitted".

henbos commented:
TODO based on Virtual Interim: remove framesEmitted from this PR (either skip it or revisit in the future with a compelling use case)

henbos commented:
Also I was asked on chat "what if the OS drops the frame?", so maybe framesCaptured should start at entry to the user agent to avoid confusion

henbos reviewed (commented):
The PR has been simplified to just count captured frames

alvestrand commented:
@youenf has questions about the getStats as a model, and will provide commentary.

youennf commented:
It is not clear to me whether we want getStats or a simple promise based getter as a method.
Are we expecting more values in getStats?

Also, it seems that we want to know whether the counter is tied to the native frame rate or the potentially decimated frame rate. If the former, we would need to know what is the actual expected native frame rate.
It is not super clear from the PR.

henbos commented:
We could make this `Promise<unsigned long long> framesCaptured()` instead if we think it is unlikely that we add more counters. My thinking was that `getStats()` could ensure everything comes from the same snapshot _if_ we do add more metrics, but I don't know if we will.

We already know the expected native frame rate, that's `track.getSettings().frameRate` (what the camera was configured to do). `framesCaptured` are the frames that were actually produced (prior to any frame decimation that the user agent may have done).

henbos commented:
E.g. I open the camera in 30 fps (settings.frameRate: 30) but due to poor lighting condition I only get 25 fps (framesCaptured would increment by 25 every second).

henbos commented:
> My understanding is that you want the post decimation counter, but this is not clear in this PR.

I was going back and forth on this, but agree that post decimation makes the most sense, so sources with different frame rates have different framesCaptured.

> Hence why I am tempted to go with a single getter.

Done. New definition:

```
                Let <var>framesCaptured</var> be a snapshot of the number of
                non-decimated captured frames, this includes frames that were
                dropped by the user agent for any reason other than decimation
                in order to achieve the configured frame rate settings.
```

PTAL

vr000m commented:
Is there a timestamp reported somewhere? 

Since this is not inheriting from getStats, we should make sure that there is a timestamp, i.e., timestamp of when the underlying object was updated, this may happen each time a framesCaptured is incremented. If a source is producing frames but it is being dropped for some reason, the timestamp will show when that started to occur.

henbos commented:
Re: @vr000m 

I figured you could just use the timestamp that the promise was resolved for this, but technically there could be a delay between the in-parallel steps and the promise resolving.

If we want a timestamp, we'd need to change this back to a stats dictionary to have multiple members.

@youennf do you have opinionos?

vr000m commented:
Let's say an issue occurred, unless the endpoint is polling framesCaptured() all the time, at least we wont know when the issue started to manifest.

henbos commented:
Friendly ping

youennf commented:
It is still a draft, @henbos, can we click 'Ready for review'?

youennf commented on index.html:
```diff
@@ -418,6 +419,81 @@ <h2>Transferable MediaStreamTrack</h2>
+                Let <var>framesCaptured</var> be a snapshot of the number of
+                non-decimated captured frames, this includes frames that were
+                dropped by the user agent for any reason other than decimation
+                in order to achieve the configured frame rate settings.
```
As I see it, the captured frames would be a value of the track's source.
Which means that if cloning a track and applying different parameters, we would still end up with the same result (modulo race conditions) when calling frameCaptured between two tracks.

Somehow, I would be tempted to add a slot to the track Source.
And then write some text explaining that on each new frame generated by the source and sent to tracks, the counter is incremented. And detail that tracks may decimate frames.

jan-ivar replied:
Sorry for the late response on this, but I think I liked it better as `getStats`. I think it's unlikely that we'll stop adding stuff (we never do 😉! I mean, timestamp?) — and it scales better (avoids dispatching for each get, which is the only cost here AFAIK). E.g.
```js
const {framesCaptured, somethingElse} = await track.getStats();
```

...is faster (1 vs N dispatches) than
```js
const framesCaptured = await track.getFramesCaptured();
const somethingElse = await track.getSomethingElse();
```

...and simpler than (no-one's gonna bother writing)
```js
const {framesCaptured, someOtherProperty} = await Promise.all([
  track.getFramesCaptured(), track.getSomethingElse()
]);
```
As for a property of the source, a track can have so many different sources, I worry about exposing a single surface for a track's source.

jan-ivar replied:
More on my reasoning: If we stay with a single get method I'd prefer it be called `getFramesCaptured` for consistency with our other get methods. And once we have a getMethod, might as well return a dictionary. That's how I arrived back at `getStats`, which in my mind would simply return a dictionary, unlike in WebRTC.

Re timestamp: would you expect a timestamp if the API were `const ms = track.framesCaptured`? If not, then it shouldn't be necessary for `const {ms} = await track.getStats()` either IMHO, because the only timing difference is that in the former case the dispatch is on some regular cadence, whereas in the latter case it's on demand. So if anything, the latter sample is going to be more more accurate to right now, not less.

henbos commented on index.html:
```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
+        <li>
+          <p>If <var>frame</var> is to be discarded by <var>track</var> in
+          order to achive the frame rate decimation of <var>track</var>'s
+          settings, abort these steps.</p>
```
Note that this wording allows the user agent to apply frame decimation earlier in the pipeline as well, since a frame not reaching the track does not trigger these in-parallel steps and that has the same behavior as aborting the steps here.

henbos commented:
I added an internal slot as suggested by @youennf and changed this back to getStats() as suggested by @jan-ivar. Can the both of you take another look?

alvestrand commented on index.html:
```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
+            <dd>
+              <p>
+                The {{timestamp}}, relative to the UNIX epoch (Jan 1, 1970, UTC),
+                for when these stats where collected.
```
This leaves the question of "on what clock" somewhat up in the air (monotonic, requires timesync...)

In RTCRtpContributingSources, we defined timestamp as Performance.timeOrigin + Performance.now() - someone claimed that Performance.now() doesn't advance while computer is suspended, so this may be the wrong definition to use, but we should strive to at least make the same errors as everyone else.

henbos replied:
I can't find anything about Performance.now() pausing while suspended, but perhaps we can simply reference the [shared monotonic clock](https://w3c.github.io/hr-time/#dfn-shared-monotonic-clock) from the High Resolution Time spec, which would remove any dependency on current document or stuff like that

alvestrand replied:
Good find. That means that if Performance.now() isn't incremented while suspended, it's a bug.

henbos replied:
OK then "Performance.timeOrigin + Performance.now()" is probably still best

alvestrand reviewed (approved):
Approving review, with a small question on timestamp definition.
I like that we're back on getStats()!

jan-ivar commented on index.html:
```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
+    <h3>MediaStreamTrack frame counters</h3>
+    <div>
+      <p>
+        Frame counters allow you to tell what the capture frame rate is, which
+        may be different from what is expected from settings - for instance, the
```
Unclear who "you" is. Might also be specific about which settings.
```suggestion
        Frame counters allow applications to tell what the capture frame rate is, which
        may be lower than the target {{MediaTrackSettings/frameRate}}  - for instance, the
```
The prose also mentions "capture" which seems to assume camera only? (camera is mentioned in a for instance, which seems ok)

jan-ivar replied:
The name itself `framesCaptured` implies this is only for video MST's returned by `getUserMedia`. Is that the intent?

If so then we have a `getStats` method that returns `{}` for other tracks, which seems fine I guess.

henbos replied:
This metric is not intended to be specific to GUM-tracks, it's just a frame counter. Do you prefer "framesProduced"? I would expect this metric to work for any track that has frames flowing

henbos replied:
Or just "frames"? They're meant to be frames not dropped, whether captured, produced through other means or from a cloned source

As a requirement it could be phrased as a counter that goes up for every non-dropped frame, measured as late in the pipeline as possible, but before the sink(s)

jan-ivar commented on index.html:
```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
+      </ol>
+    </div>
+    <div>
+      <p>Monitor frames that arrive to <var>track</var> from its source by
+      running the following in-parallel steps for every <var>frame</var>:</p>
+      <ol>
+        <li>
+          <p>If <var>frame</var> is to be discarded by <var>track</var> in
+          order to achive the frame rate decimation of <var>track</var>'s
+          settings, abort these steps.</p>
```
I hesitate introducing algorithms here, since this seems deep in implementation territory.

Algorithms are nice when dealing with JS-observable operations, but it's not JS-observable when frames "arrive to" to a MediaStreamTrack, whatever that means.

This might be better expressed as requirements.

Again, linking to the frameRate constraint may be helpful (if this feature is tied to it existing).

As [noted](https://w3c.github.io/mediacapture-main/#media-track-supported-constraints) in the spec: _"The constraints specified in this specification apply only to instances of [MediaStreamTrack](https://w3c.github.io/mediacapture-main/#dom-mediastreamtrack) generated by [MediaDevices](https://w3c.github.io/mediacapture-main/#dom-mediadevices).[getUserMedia](https://w3c.github.io/mediacapture-main/#dom-mediadevices-getusermedia)(), unless stated otherwise in other specifications."_

henbos replied:
> I hesitate introducing algorithms here, ...

> This might be better expressed as requirements.

I tried without an algorithm but there as confusion, then Youenn suggested an internal slot to make things less confusing. Considering the amount of back and forth, I don't think this is a bad idea. But the important thing is everyone is on the same page with what we're doing.

> ...it's not JS-observable when frames "arrive to" a MediaStreamTrack, whatever that means.

It's not a JS-observable event, which should give implementations some amount of wiggle-room. I'm just trying to count frames at a sensible time. How would this be phrased otherwise?

Note that there is a difference between counting in the capture processes or renderer process etc, hence the "arrive at the track" language, even though that is not the final destination.

henbos replied:
I changed the PR to be requirements-based

jan-ivar commented on index.html:
```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
+            </li>
+            <li>
+              <p>
+                In parallel, begin the [= in-parallel steps to get track
+                stats =], given <var>track</var> and <var>p</var>.
```
Normally we just inline in-parallel steps. Any reason not to do that here as well?

jan-ivar commented on index.html:
```diff
@@ -418,6 +419,145 @@ <h2>Transferable MediaStreamTrack</h2>
+          <ol>
+            <li>
+              <p>
+                Let <var>timestamp</var> be the current time relative to the
+                UNIX epoch (Jan 1, 1970, UTC).
```
Is there any precedent for an API returning a timestamp that is basically now + the time it takes to queue a task on main thread? Seems of low value to me. What is the resolution of this timer? The only time it would differ substantially is if the main thread is under heavy load that tasks have already been queued ahead of of this task. 

To mitigate this, since the read is "atomic", have we considered reading it in the queued task? This would remove any need for a timestamp IMHO.

henbos replied:
RTCPeerConnection.getStats() is an example of using timestamp for the same reason, but if we can make an atomic read from the main thread then we can probably get rid of timestamp, but then the UA would no longer be allowed to implement this using a single threaded frame counter off the main thread, which I think would be nice to be allowed to do

henbos replied:
I changed the PR to be requirements-based and not have internal slots or "atomic" language, so now the timestamp is needed to avoid forcing how to implement this

jan-ivar reviewed (commented):
Patch should be rewritten to not assume all tracks come from camera.

To merge this, do we want to consider a separate issue on the timestamp?

jan-ivar commented:
We might also want to put the final API in front of the working group before merging.

henbos commented:
I updated the PR again. I still have timestamp in the same PR, but I clarified that this is any video track and tried to phrase it more like a requirement. I also found another spec which has more requirement-like phrasing: https://w3c.github.io/media-playback-quality/#videoplaybackquality-interface

Please take another look and tell me if we're getting any closer to merging this

vr000m commented:
Is this on the agenda for Jan 2023 interim?

henbos commented:
I can create a slide for it

henbos commented:
Virtual Interim feedback:
- Overall approach of requirements and categories makes sense, but...
- "Total frames" makes more sense than "dropped".
- We need to clarify what happens while muted (proposal: don't increment any counters).

dontcallmedom commented:
> "Total frames" makes more sense than "dropped".

I like that this also allows to add more fine-grained categories over time if we find that more diagnostics would be useful

henbos reviewed (commented):
Please take another look

henbos commented:
I added a note about the total frames and calculating lost frames from total frames. Since this is only a requirement and not an algorithm, and frames could get lost anywhere including outside the renderer process so I noted that this is best effort

henbos commented:
I know I said that when the track is muted then the frame counters should stop incrementing, but at the time I was thinking about camera frames. @handellm reminded me that a muted track in fact produces zero content frames which also get delivered, so I updated the PR to reflect that total frames may in fact increment even while muted

youennf commented:
> @handellm reminded me that a muted track in fact produces zero content frames which also get delivered

In WebKit at least, a muted track produces no frame at all. It would be nice if we could be consistent across UAs.
RTCPeerConnection will send encoded zero content frames periodically if tracks are muted but this is a different mechanism.

Can you validate Chrome behaviour here? Say with VideoTrackProcessor or rvfc maybe?

guidou commented:
> > @handellm reminded me that a muted track in fact produces zero content frames which also get delivered
> 
> In WebKit at least, a muted track produces no frame at all. It would be nice if we could be consistent across UAs. RTCPeerConnection will send encoded zero content frames periodically if tracks are muted but this is a different mechanism.
> 
> Can you validate Chrome behaviour here? Say with VideoTrackProcessor or rvfc maybe?

Muted tracks in Chrome don't produce frames either.

henbos commented:
Oh... so when the spec says "A muted or disabled [MediaStreamTrack](https://rawgit.com/w3c/mediacapture-main/master/getusermedia.html#dom-mediastreamtrack) renders either silence (audio), black frames (video), or a zero-information-content equivalent" this refers to what the sink is doing, not what the track's source is doing?

henbos commented:
I updated the PR to instead say "If no frames are flowing, such as if the track is muted or disabled, then total frames does not increment."

henbos commented:
> I updated the PR to instead say "If no frames are flowing, such as if the track is muted or disabled, then total frames does not increment."

If "frames flowing" is an implementation detail, I could rephrase this to simply "If the track is muted or disabled, then total frames does not increment."

vr000m commented:
Depending on which type of mute is implemented, source will send encrypted 0 content frames (silent samples) -- I think it's done for security reasons within the peerconnection. 

However, in this case, if you mute the source, then the source isn't going to produce any frames.

henbos commented:
I think the current phrasing is good. Can we merge this PR soon?

alvestrand commented:
Varun, are you OK with merging?
What a peerconnection does with the frames is a webrtc matter, not a mediacapture matter.

vr000m commented:
Agree on the pc being a matter for another spec. I've followed along the conversation in this PR but haven't looked at the latest changes. 

Give me a day to review the latest changes.

vr000m commented on index.html:
```diff
@@ -418,6 +419,163 @@ <h2>Transferable MediaStreamTrack</h2>
+          reason. The number of dropped frames can be calculated by subtracting
+          [= delivered frames =] and [= discarded frames =] from
+          [= total frames =].</p>
+          <div class="note">
```
We haven't defined what dropped means. It seems to me that dropped is a catch-all category that is anything that's not categorized as delivered or discarded.

henbos replied:
But it is meant to be a catch-all category for anything that is not categorized as delivered or discarded, so it seems the definition fits the bill.

If we want to figure out why a frame was lost we could add more counters later, but different user agents may drop for different reasons. The interesting thing about total frames is that it tells you the camera frame rate, which is not measurable without this metric.

vr000m replied:
perhaps:

_The number of dropped frames for various unknown reasons can be calculated by subtracting_

henbos replied:
Done

vr000m reviewed (approved):
No real blockers. Left a comment on discarded in the review. Perhaps we can clarify that discarded frames are discarded for unknown reasons

youennf commented on index.html:
```diff
@@ -418,6 +419,163 @@ <h2>Transferable MediaStreamTrack</h2>
+    <h3>MediaStreamTrack frame stats</h3>
+    <div>
+      <p>
+        On any track that supports {{MediaTrackSettings/frameRate}} as a
```
This seems to indicate that peer connection tracks would have this counter.
Does this make sense? It seems we mostly care about cameras and screen share.
Can we find wording to do this?

henbos replied:
Done. I reworded this as "is sourced from getUserMedia() or getDisplayMedia()"

youennf replied:
OK, this seems good enough as a first iteration.
In the future, we might want to add a [hasFrameStats] slot to the track's source.
If the slot is not there or is false, we would reject getFrameStats().
mediacapture-main and mediacapture-screenshare specs would then set this slot to true.

henbos replied:
Good suggestion for the future! Then we can let each individual spec decide.

alvestrand commented:
So if the question gets answered, Youenn can give an approval, and then we can merge it. There seems to be no other issues.

henbos commented on index.html:
```diff
@@ -418,6 +419,163 @@ <h2>Transferable MediaStreamTrack</h2>
+        {{MediaTrackSettings/frameRate}}. For example, if the track is sourced
+        from a camera then the production of frames could be slowed down if it's
+        dark or frames could be dropped if the system is CPU starved. This could
+        impact the total number of frames handled by the source and impact how
```
Done

henbos commented on index.html:
```diff
@@ -418,6 +419,163 @@ <h2>Transferable MediaStreamTrack</h2>
+    </div>
+    <div>
+      <p>If a {{MediaStreamTrack}} is sourced from <code>getUserMedia()</code>
+      or <code>getDisplayMedia()</code>, the user agent is required to count
```
Note: I was unable to get `{{MediaDevices/getDisplayMedia()}}` to link correctly (despite by best attempt at updating the terminology section) so I used `<code>getDisplayMedia()</code>` instead. Unless someone knows how to fix this, I suggest we either create a follow-up issue "Fix getDisplayMedia() reference" or that I rephrase this as "sourced from a camera or screenshare"

henbos commented:
The comments have been addressed, PTAL
